### PR TITLE
Issue #3516: Prepare system

### PR DESCRIPTION
**Pull request by simnandez**

### CODE REVIEW

File: htdocs/admin/user.php [hunk 1]
```diff
@@ -136,6 +136,26 @@
...
+    $var = !$var;
+    print '<tr ' . $bc[$var] . '>';
+    print '<td>' . $langs->trans("UserUsesHierarchy") . '</td>';
+    print '<td align="center" width="20">&nbsp;</td>';
+
+    print '<td align="center" width="100">';
+    if ($conf->use_javascript_ajax) {
+        print ajax_constantonoff('USERS_USE_HIERARCHY');
+    }
+    else {
```
Comment (rdoursenaud) [inline htdocs/admin/user.php:150]: PSR-1 says this should be on the same line as the previous closing brace ;)

File: htdocs/user/class/user.class.php [hunk 2]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+	 */
+	private function get_childrens($userid)
+	{
+		$sql = "SELECT rowid FROM ".MAIN_DB_PREFIX."user";
+		$sql.= " WHERE fk_user = ".$userid;
+
+		dol_syslog(__METHOD__."  sql=".$sql, LOG_DEBUG);
+		$res  = $this->db->query($sql);
+		if ($res)
+		{
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2315]: Same here.

File: htdocs/user/class/user.class.php [hunk 3]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+			while ($rec = $this->db->fetch_array($res))
+			{
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2317]: and here

File: htdocs/user/class/user.class.php [hunk 4]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+				$this->childrens[] = $rec['rowid'];
+				$this->get_childrens($rec['rowid']);
+			}
+		}
+		else
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2322]: And there

File: htdocs/user/class/user.class.php [hunk 5]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+			$hierarchy = $this->childrens;
+
+			$max = count($hierarchy);
+			if ($max) {
+				$users .= "( $this->id,";
+				foreach ($hierarchy as $i => $value) {
+					if (($i + 1) < $max) {
+						$users .= $value . ",";
+					}
+					else {
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2351]: And here

File: htdocs/user/class/user.class.php [hunk 6]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+					}
+				}
+			}
+
+			if (!$users) {
+				$users = "(" . $this->id . ")";
+			}
+
+		}
+		else {
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2362]: And there again

File: htdocs/user/class/user.class.php [hunk 7]
```diff
@@ -2302,6 +2304,74 @@ function get_children()
...
+
+	/**
+	 * @param $userid User to get hierarchy
+	 *
+	 * @return string of user for SQL
+	 */
+	function get_hierarchy($userid)
+	{
+		global $conf;
+		$users = "";
```
Comment (rdoursenaud) [inline htdocs/user/class/user.class.php:2337]: Wouldn't an array be better suited than a string?

**Comment by eldy** [comment]
What is the goal of introducing an option USERS_USE_HIERARCHY ?
If a user don't want to have hierarchy define, he just have to not enter the supervisor.

**Comment by simnandez** [comment]
The goal of this is enable or not the hierarchy throughout the system. ex. if I am below you in the user hearchy and we active USERS_USE_HIERARCHY yo can view my thirds, invoices...

**Comment by eldy** [comment]
Currently visibility already depends on permissions and link with sale representative or not and if user is external or not. Introducing a fourth condition coming from a 4th origin will make permission rules really difficult to understand.
This should not be an option (even hidden) but a permission. And because it is feature that will probably be in conflict with other permission rules, it should be a hidden permission.
So hidden option "USERS_USE_HIERARCHY", should just enable visibility of the permission "Read thirdparty my supervisor is able to read"

**Comment by eldy** [comment]
Closed too old no valid PR

**Pull request closed without merging**